EagerExecution.swift.gyb

```diff
@@ -0,0 +1,314 @@
+        }
+      }
+      dims.withUnsafeMutableBufferPointer { dimsBuffer in
+        ranks.withUnsafeBufferPointer { ranksBuffer in
+          TFE_OpSetAttrShapeList(
+            op, name, dimsBuffer.baseAddress, ranksBuffer.baseAddress,
+            Int32(ranksBuffer.count), status)
+        }
+      }
+    }
+  }
+
+  @inlinable @inline(__always)
+  internal func setAttr(_ name: String, _ value: [TensorShape?]) {
+    let flattenedDims = value.flatMap { (tensorShapeOpt) -> [Int64] in
+      if let tensorShape = tensorShapeOpt {
+        return tensorShape.dimensions.map(Int64.init)
+      }
+      return []
+    }
+    let ranks = value.map { shape in (shape?.rank).map(Int32.init) ?? -1 }
+    flattenedDims.withUnsafeBufferPointer { flattenedDimsBuffer in
+      var dimsPtr: UnsafePointer<Int64>? = flattenedDimsBuffer.baseAddress
+      var dims: [UnsafePointer<Int64>?] = []
+      for rank in ranks {
+        dims.append(dimsPtr)
+        if rank >= 0 {
+          dimsPtr = dimsPtr.map { $0.advanced(by: Int(rank)) }
+        }
+      }
+      dims.withUnsafeMutableBufferPointer { dimsBuffer in
+        ranks.withUnsafeBufferPointer { ranksBuffer in
+          TFE_OpSetAttrShapeList(
+            op, name, dimsBuffer.baseAddress, ranksBuffer.baseAddress,
+            Int32(ranksBuffer.count), status)
+        }
+      }
+    }
+  }
+
+  @inlinable @inline(__always)
+  internal func setAttr<In: TensorGroup, Out: TensorGroup>(_ name: String, _ value: (In) -> Out) {
+    _tffunc(value).utf8CString.withUnsafeBufferPointer { buffer in
+      // utf8CString is null-terminated; TFE_OpSetAttrFunctionName wants
+      // non-null-terminated.
+      TFE_OpSetAttrFunctionName(op, name, buffer.baseAddress, buffer.count - 1)
+    }
+  }
+
+  /// **WARNING:** After constructing a `TFE_Op`, any one of its `execute` methods must be called
+  /// *exactly once*. If not called, then a memory leak is introduced due to the underlying
+  /// TensorFlow eager op object not being freed. If called more than once, then a SEGFAULT may
+  /// occur due to trying to execute a TensorFlow eager op that has already been freed.
+
+  @inlinable @inline(__always)
+  internal func execute() {
+    var count: Int32 = 0
+    var unused: CTensorHandle?
+    _TFCEagerExecute(op, &unused, &count, status)
+    checkOk(status)
+    TFE_DeleteOp(op)
+    TF_DeleteStatus(status)
+  }
+
+%for n in range(1, 10 + 1):
+  @inlinable @inline(__always)
+  internal func execute<${", ".join(["T" + str(i) + " : TensorArrayProtocol" for i in range(n)])}>(
+    ${",\n    ".join(["_ count" + str(i) + ": Int" for i in range(n)])}
+  ) -> (${", ".join(["T" + str(i) for i in range(n)])}) {
+    var count = ${" + ".join(["Int32(count" + str(i) + ")" for i in range(n)])}
+    let buffer: UnsafeMutablePointer<CTensorHandle> =
+      UnsafeMutablePointer.allocate(capacity: Int(count))
+    _TFCEagerExecute(op, UnsafeMutablePointer<CTensorHandle?>(buffer), &count, status)
```

Comment:
This is also emitted by the tfop emission process:
_TFCOpSetDeviceFromScope(op, status)

Without it, multi-device support is broken.